src/cargo/ops/cargo_package.rs

```diff
@@ -231,9 +212,33 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option
 fn build_ar_list(
     ws: &Workspace<'_>,
     pkg: &Package,
-    src_files: Vec<PathBuf>,
-    vcs_info: Option<VcsInfo>,
+    opts: &PackageOpts<'_>,
 ) -> CargoResult<Vec<ArchiveFile>> {
+    let gctx = ws.gctx();
+    let mut src = PathSource::new(pkg.root(), pkg.package_id().source_id(), gctx);
+    src.update()?;
+
+    if opts.check_metadata {
+        check_metadata(pkg, gctx)?;
+    }
+
+    if !pkg.manifest().exclude().is_empty() && !pkg.manifest().include().is_empty() {
+        gctx.shell().warn(
+            "both package.include and package.exclude are specified; \
+             the exclude list will be ignored",
+        )?;
+    }
+    let src_files = src.list_files(pkg)?;
+
+    // Check (git) repository state, getting the current commit hash if not
+    // dirty.
+    let vcs_info = if !opts.allow_dirty {
+        // This will error if a dirty repo is found.
+        check_repo_state(pkg, &src_files, gctx)?
+    } else {
+        None
+    };
```

Comment:
metadata seems unrelated to this step

I feel like repo state is similar. If this function read the files also, then I could see that.